Libs/libMacGitverCore/Widgets/ProgressDlg.cpp

```diff
@@ -0,0 +1,261 @@
+{
+    if (mUpdater.isActive()) {
+        // restart timer with the new interval
+        mUpdater.start(msec);
+    }
+    else {
+        mUpdater.setInterval(msec);
+    }
+}
+
+void ProgressDlg::addActivity(const QString& description, QObject* activity,
+                              const StepInfo::List& steps)
+{
+    Q_ASSERT(activity);
+
+    Private::ProgressWdgt* a = new Private::ProgressWdgt(description);
+    mActivities[activity] = a;
+
+    QTreeWidgetItem* activityItem(new QTreeWidgetItem);
+    ui->treeProgress->addTopLevelItem(activityItem);
+    ui->treeProgress->setItemWidget(activityItem, 0, a);
+
+    foreach (const StepInfo& sir, steps) {
+        Private::ProgressWdgt* s = new Private::ProgressWdgt(sir.desc);
+        a->mSteps[sir.name] = s;
+
+        QTreeWidgetItem* stepItem(new QTreeWidgetItem);
+        activityItem->addChild(stepItem);
+        ui->treeProgress->setItemWidget(stepItem, 0, s);
+    }
+
+    a->txtStatusInfo->setText(tr("Running..."));
+}
+
+void ProgressDlg::setStatusInfo(QObject* activity, const QString& step,
+                                const QString& text)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->txtStatusInfo->setText(text);
+}
+
+void ProgressDlg::setPercentage(QObject* activity, const QString& step,
+                                qreal percent)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->mPercentage = qMin( qMax(percent, 0.), 1. ) * 100.;
+}
+
+void ProgressDlg::closeEvent( QCloseEvent* ev )
+{
+    if( !mDone )
+    {
+        ev->ignore();
+        return;
+    }
+
+    QDialog::closeEvent( ev );
+}
+
+void ProgressDlg::finished(QObject* activity)
+{
+    Private::ProgressWdgt* a = mActivities[activity];
+    Q_ASSERT(a);
+
+    a->mActive = false;
+    a->txtStatusInfo->setText(tr("Finished!"));
+    a->resultChanged(Private::ProgressWdgt::Result::Ok);
+}
+
+void ProgressDlg::finished(QObject* activity, const QString& step)
+{
+    Private::ProgressWdgt* s = findStep(activity, step);
+    Q_ASSERT(s);
+
+    s->mActive = false;
+    s->mPercentage = 100.;
+    s->resultChanged(Private::ProgressWdgt::Result::Ok);
+}
+
+void ProgressDlg::setError(QObject* activity, const QString& message)
+{
+    Private::ProgressWdgt* a = mActivities[activity];
+    Q_ASSERT(a);
+    a->mActive = false;
+    a->txtStatusInfo->setText(tr("Failed!"));
+    a->setToolTip(tr("<b>Activity failed!</b><hr/>%1").arg(message));
+    a->resultChanged(Private::ProgressWdgt::Result::Error);
+}
+
+void ProgressDlg::updateActivities()
+{
+    Activities::Iterator it = mActivities.begin();
+    while ( it != mActivities.end() ) {
+        Private::ProgressWdgt* a = *it;
+        a->mPercentage = 0.;
+
+        foreach (Private::ProgressWdgt* s, a->mSteps) {
+            s->progressBar->setValue(qRound(s->mPercentage));
+            qreal stepPercent = s->mPercentage / a->mSteps.size();
```

Comment:
Not in my book. However, once `libActivities` is done, you won't need to calculate that data since it has it ready for you (and with a slightly different algorithm).